dockers/texcmp/texcmp.js

```diff
@@ -0,0 +1,244 @@
+    var diffFile = path.join(diffDir, key + ".png");
+
+    // Step 1: write key.tex file
+    var fftLatex = Q.nfcall(fs.writeFile, texFile, tex).then(function() {
+        // Step 2: call "pdflatex key" to create key.pdf
+        return execFile("pdflatex", [
+            "-interaction", "nonstopmode", key
+        ], {cwd: tmpDir});
+    }).then(function() {
+        console.log("Typeset " + key);
+        // Step 3: call "convert ... key.pdf key.png" to create key.png
+        return execFile("convert", [
+            "-density", dpi, "-units", "PixelsPerInch", "-flatten",
+            pdfFile, pngFile
+        ]);
+    }).then(function() {
+        console.log("Rasterized " + key);
+        // Step 4: apply FFT to that
+        return fftPNG(pngFile);
+    });
+    // Step 5: apply FFT to reference image as well
+    var fftBrowser = fftPNG(browserFile);
+
+    return Q.all([fftBrowser, fftLatex]).spread(function(browser, latex) {
+        // Now we have the FFT result from both 
+        // Step 6: find alignment which maximizes overlap.
+        // This uses a FFT-based correlation comoputation.
+        var x, y;
+        var real = newMatrix();
+        var imag = newMatrix();
+
+        // Step 6a: (real + i*imag) = latex * conjugate(browser)
+        for (y = 0; y < alignHeight; ++y) {
+            for (x = 0; x < alignWidth; ++x) {
+                var br = browser.real.get(y, x);
+                var bi = browser.imag.get(y, x);
+                var lr = latex.real.get(y, x);
+                var li = latex.imag.get(y, x);
+                real.set(y, x, br * lr + bi * li);
+                imag.set(y, x, br * li - bi * lr);
+            }
+        }
+
+        // Step 6b: (real + i*imag) = inverseFFT(real + i*imag)
+        fft(-1, real, imag);
+
+        // Step 6c: find position where the (squared) absolute value is maximal
+        var ox = 0;
+        var oy = 0;
+        var ov = -1;
+        for (y = 0; y < alignHeight; ++y) {
+            for (x = 0; x < alignWidth; ++x) {
+                var or = real.get(y, x);
+                var oi = imag.get(y, x);
+                var sq = or * or + oi * oi;
+                if (ov < sq) {
+                    ov = sq;
+                    ox = x;
+                    oy = y;
+                }
+            }
+        }
+
+        // Step 6d: Treat negative offsets in a non-cyclic way
+        if (oy > (alignHeight >>> 1)) {
+            oy -= alignHeight;
+        }
+        if (ox > (alignWidth >>> 1)) {
+            ox -= alignWidth;
+        }
+        console.log("Positioned " + key + ": " + ox + ", " + oy);
+
+        // Step 7: use these offsets to compute difference illustration
+        var uw = Math.max(browser.width, latex.width) + Math.abs(ox);
+        var uh = Math.max(browser.height, latex.height) + Math.abs(oy);
+        var bx = Math.max(ox, 0);
+        var by = Math.max(oy, 0);
+        var lx = Math.max(-ox, 0);
+        var ly = Math.max(-oy, 0);
```

Comment:
Can I add comments instead? I'd prefer to keep the `-extent` arguments in one line.